### PR TITLE
Handle bind-mounts and use devicemapper device when available

### PR DESCRIPTION
**Pull request by amshinde**
This PR handles the mounts passed with the OCI spec file. It handles the bind-mounts by passing them to the VM through 9pfs and handling the bind-mounting within the VM to work for both cases where we plan to use overlay as rootfs or a block device.
When devicemapper is the storage driver for the container rootfs, use the underlying block device instead.

### CODE REVIEW

**Comment by amshinde** [comment]
This PR requires hyperstart PR to be merged first and made available in the container image.
https://github.com/clearcontainers/hyperstart/pull/32

File: filesystem.go [hunk 1]
```diff
@@ -570,6 +599,32 @@ func (fs *filesystem) storeContainerProcess(podID, containerID string, process P
...
+	if err != nil {
+		return []Mount{}, err
+	}
+
+	switch mounts := data.(type) {
+	case []Mount:
+		return mounts, nil
+	}
+
+	return []Mount{}, fmt.Errorf("Unknown mounts type")
```
Comment (jodh-intel) [inline filesystem.go]: Could you add the type (`%T`) to the error?

File: filesystem_test.go [hunk 2]
```diff
@@ -325,6 +326,79 @@ func TestFilesystemFetchContainerConfigFailingPodIDEmpty(t *testing.T) {
...
+	}
+
+	source := "/dev/sda1"
+	dest := "/root"
+	mntType := "ext4"
+	options := "rw"
+	ignoreMount := false
+	hostPath := "/tmp/root"
+
+	mountData := fmt.Sprintf("[{\"Source\":\"%s\",\"Destination\":\"%s\",\"Type\":\"%s\",\"Options\": [\"%s\"],\"IgnoreMount\":%s, \"HostPath\":\"%s\"}]", source, dest, mntType, options, strconv.FormatBool(ignoreMount), hostPath)
```
Comment (jodh-intel) [inline filesystem_test.go]: This would look better and be easier to read if you used backticks:

```
mountData := fmt.Sprintf(`
  [
     {
        "Source": "%s",
  ...
`, source, ...)

File: mount.go [hunk 3]
```diff
@@ -16,5 +16,78 @@
...
+	// stat.Dev points to the underlying device containing the file
+	major := major(stat.Dev)
+	minor := minor(stat.Dev)
+
+	mountPoint := path
+
+	// We get the mount point by recursively peforming stat on the path
+	// The point where the device changes indicates the mountpoint
+	for {
+		if mountPoint == "/" {
```
Comment (mcastelino) [inline mount.go]: We should still check if / results in a switch. If it does not, then we do not have a mount point.

File: mount.go [hunk 4]
```diff
@@ -91,3 +95,43 @@ func getDeviceForPath(path string) (*device, error) {
...
+		if err == io.EOF {
+			break
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) != fieldsPerLine {
+			return "", "", fmt.Errorf("Incorrect no of fields (expected %d, got %d)) :%s", fieldsPerLine, len(fields), line)
+		}
+
+		// fields[0] => Device, fields[1] => Path , fields[2] => Type
```
Comment (mcastelino) [inline mount.go]: would having enums here help vs 0, 1, 2? You already have a constant definition above

File: mount.go [hunk 5]
```diff
@@ -91,3 +95,43 @@ func getDeviceForPath(path string) (*device, error) {
...
 	return dev, nil
 }
+
+const (
+	procMountsFile = "/proc/mounts"
+
+	fieldsPerLine = 6
+)
+
+func getDevicePathAndFsType(mountPoint string) (string, string, error) {
```
Comment (mcastelino) [inline mount.go]: Nit, named return values would help here.

File: hyperstart.go [hunk 6]
```diff
@@ -437,6 +481,10 @@ func (h *hyper) startPod(pod Pod) error {
 		}
 	}
 
+	if err := h.addDrives(&pod); err != nil {
```
Comment (mcastelino) [inline hyperstart.go]: Do you still pass the mounted devicemapper mount pass via 9pfs? I did not see code to disable that? Did I miss it. I am worried about block device corruption if we continue to pass it in via 9p. Even thing like accidental atime update

Also is there a easy way to toggle the behaviour. i.e, switch back to 9pfs even for devicemapper. Offering this option may be required if there are users who may care about the lack of support for dynamic file copy in/out of the container via 

```
docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH|-
docker cp [OPTIONS] SRC_PATH|- CONTAINER:DEST_PATH
```

Also we need to document this limitation with the direct assignment approach.

File: pkg/oci/utils_test.go [hunk 7]
```diff
@@ -86,18 +86,24 @@ func TestMinimalPodConfig(t *testing.T) {
...
 		},
 		{
 			Source:      "devpts",
 			Destination: "/dev/pts",
 			Type:        "devpts",
 			Options:     []string{"nosuid", "noexec", "newinstance", "ptmxmode=0666", "mode=0620", "gid=5"},
+			IgnoreMount: false,
+			HostPath:    "",
 		},
 	}
```
Comment (amshinde) [inline pkg/oci/utils_test.go]: @mcastelino  I though /dev/shm needs to be created inside the VM, is that right. I am planning to send all the mount info for the mounts that need to be created within the VM (including /dev/shm) so that hyperstart can create those. Planning to send a separate PR for this.(for virtcontainers and hyperstart)

File: hyperstart.go [hunk 8]
```diff
@@ -220,6 +222,79 @@ func (h *hyper) bindMountContainerRootfs(podID, cID, cRootFs string, readonly bo
...
+		}
+	}
+	return false
+}
+
+// bindMountContainerMounts handles bind-mounts by bindmounting to the host shared directory
+// which is mounted through 9pfs in the VM.
+// Hyperstart uses "fsmap" struct to bind mount bind these mounts in the hypertsart shared directory
+// to the correct mountpoint within the container rootfs.
+func (h *hyper) bindMountContainerMounts(podID string, c Container, mounts *[]Mount) ([]*hyperstart.FsmapDescriptor, error) {
```
Comment (amshinde) [inline hyperstart.go]: Yes, I dont need a pointer here. Will fix this.

File: filesystem_test.go (hunk 2 above)
Comment (amshinde) [inline filesystem_test.go]: Didnt know about the backticks. Will fix this.

File: filesystem.go (hunk 1 above)
Comment (amshinde) [inline filesystem.go]: Fixing this

File: mount.go (hunk 5 above)
Comment (amshinde) [inline mount.go]: Fixing.

File: pkg/oci/utils_test.go (hunk 7 above)
Comment (mcastelino) [inline pkg/oci/utils_test.go]: Yes. It has to be created inside. But the size needs to be picked up from the outside. So I thought sending this value in via the same mechanism would make sense.

File: hyperstart.go (hunk 6 above)
Comment (amshinde) [inline hyperstart.go]: Yes, we are still passing the mounted devicemapper mount pass via 9pfs. I didnt think this would cause any issues. I'll disable the 9pfs mount.
Will add the documentation too.
I suppose we could add a configuration option in  the global config file for the runtime to toggle the behaviour. I'll add that to the runtime.

File: pkg/oci/utils_test.go (hunk 7 above)
Comment (amshinde) [inline pkg/oci/utils_test.go]: Yes, I was planning to send this is a separate structure say "mounts" to hyperstart. But I suppose i could get away with extending the current fsmap type to handle this as well. I'll add the system mounts handling in a separate PR.

File: hyperstart.go [hunk 9]
```diff
@@ -220,6 +222,79 @@ func (h *hyper) bindMountContainerRootfs(podID, cID, cRootFs string, readonly bo
...
+	var fsmap []*hyperstart.FsmapDescriptor
+
+	// TODO: We need to handle system mounts by having the agent create them inside the VM.
+	// Handle just bind mounts for now
+
+	for ind := range *mounts {
+		m := &(*mounts)[ind]
+
+		m.IgnoreMount = true
+		if isPartOf(systemMounts, m.Destination) {
```
Comment (amshinde) [inline hyperstart.go]: Yes makes sense.

File: pkg/oci/utils_test.go [hunk 10]
```diff
@@ -80,6 +80,27 @@ func TestMinimalPodConfig(t *testing.T) {
...
 
+	expectedMounts := []vc.Mount{
+		{
+			Source:      "proc",
+			Destination: "/proc",
+			Type:        "proc",
+			Options:     nil,
+		},
+		{
```
Comment (amshinde) [inline pkg/oci/utils_test.go]: @mcastelino  These are just the tests, but I am currently doing the bind-mounts in the same namespace as host. Need to add the change to enter a separate mount namespace for 3.0

File: mount.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+// These mounts need to be created by the agent within the VM
+var systemMounts = []string{"/proc", "/dev", "/dev/pts", "/dev/shm", "/dev/mqueue", "/sys", "/sys/fs/cgroup"}
```
Comment (amshinde) [inline mount.go]: The /dev would have been sufficient here, basically I ignore mounts with that prefix.We use bind inside the VM. I think the fsmap structure needs to be extended to include extra mount options.

File: mount.go (hunk 4 above)
Comment (amshinde) [inline mount.go]: fixing

File: mount.go (hunk 3 above)
Comment (amshinde) [inline mount.go]: @mcastelino I dont quite understand what you mean, I am using "/" as the termination case while walking up the directory tree. When i reach "/", that should be the mount point.

Comment (mcastelino) [inline mount.go]: @amshinde let us take two cases.

If you have /mydisk where mydisk is just a directory on / 
vs /mydisk where you have mounted something on /mydisk

In the first case /mydisk and / will have the same stat
In the second case /mydisk and / will have different stats.

So in the first case we error out saying we did not find a mount point.
In the second case /mydisk is the mount point.

The only time / is a valid mount point will be in the case of volumes where the user may have done something along the lines of -v /:/hostdisk.
In such a case yes / is the mount point. But that is a special case, where the source is /.

File: hyperstart.go [hunk 12]
```diff
@@ -292,6 +292,7 @@ func (h *hyper) bindMountContainerMounts(podID string, c Container, mounts []Mou
 		fsmap = append(fsmap, fsmapDesc)
 	}
 
+	c.storeMounts()
```
Comment (sameo) [inline hyperstart.go]: The container mount storage is generic operation and should be done at the container or pod level. We don't want each potential hypervisor interface implementation to explicitly call this one for each container it creates.

File: hyperstart.go [hunk 13]
```diff
@@ -296,6 +296,24 @@ func (h *hyper) bindMountContainerMounts(podID string, c Container, mounts []Mou
 	return fsmap, nil
 }
 
+func (h *hyper) bindUnmountContainerMounts(podID string, c *Container) error {
```
Comment (sameo) [inline hyperstart.go]: This is a container generic routine and should be implemented as a private `Container` method instead. Again, we want that one to be usable by all potential hypervisor implementations.

File: hyperstart.go [hunk 14]
```diff
@@ -305,7 +323,9 @@ func (h *hyper) bindUnmountContainerRootfs(podID, cID string) error {
 
 func (h *hyper) bindUnmountAllRootfs(pod Pod) {
 	for _, c := range pod.containers {
+		h.bindUnmountContainerMounts(pod.id, c)
 		h.bindUnmountContainerRootfs(pod.id, c.id)
+		h.bindUnmountContainerMounts(pod.id, c)
```
Comment (sameo) [inline hyperstart.go]: Why the double unmount ?

File: pod.go [hunk 15]
```diff
@@ -212,6 +212,14 @@ func (s *Sockets) String() string {
 	return strings.Join(sockSlice, " ")
 }
 
+// Drive represents a block storage drive in the case where the graph storage
+// driver has an underlying block storage device.
+type Drive struct {
```
Comment (sameo) [inline pod.go]: You may want to document this structure fields.

File: mount.go [hunk 16]
```diff
@@ -16,5 +16,78 @@
...
+
+func major(dev uint64) int {
+	return int((dev >> 8) & 0xfff)
+
+}
+
+func minor(dev uint64) int {
+	return int((dev & 0xff) | ((dev >> 12) & 0xfff00))
+}
+
```
Comment (sameo) [inline mount.go]: Something like:

```Go
var (	
        MINORBITS = uint(20)
        MINORMASK = uint((uint(1) << MINORBITS) - 1)
)

func major(dev uint) int{
	return int(dev >> MINORBITS)
}

func minor(dev uint) int{
	return int(dev & MINORMASK)
}

```

would be simpler.

File: mount.go [hunk 17]
```diff
@@ -16,5 +16,78 @@
...
+type device struct {
+	major      int
+	minor      int
+	mountPoint string
+}
+
+// getDeviceForPath gets the underlying device containing the file specified by path.
+// The device type constitutes the major-minor number of the device and the dest mountPoint for the device
+func getDeviceForPath(path string) (*device, error) {
```
Comment (sameo) [inline mount.go]: Please just return a `device`

File: mount.go [hunk 18]
```diff
@@ -16,5 +16,78 @@
...
+}
+
+type device struct {
+	major      int
+	minor      int
+	mountPoint string
+}
+
+// getDeviceForPath gets the underlying device containing the file specified by path.
+// The device type constitutes the major-minor number of the device and the dest mountPoint for the device
```
Comment (sameo) [inline mount.go]: Could you give an example of what this function is supposed to do ? Given path `/foo/bar/` what would be the device ?
For example, let's say that `/dev/sda1` is mounted on `/a/b/c/`. If I pass `/a/b/c/file1` to this routine it should return:
```Go
device{
     major: major(/dev/sda3)
     minor: minor(/dev/sda3)
     mountPoint: /a/b/c/
}
```

Correct ?

File: mount.go [hunk 19]
```diff
@@ -91,3 +95,56 @@ func getDeviceForPath(path string) (*device, error) {
...
+	fieldsPerLine = 6
+)
+
+const (
+	procDeviceIndex = iota
+	procPathIndex
+	procTypeIndex
+)
+
+func getDevicePathAndFsType(mountPoint string) (devicePath, fsType string, err error) {
```
Comment (sameo) [inline mount.go]: So there's a kubernetes package for doing a bunch of mount and device operations:
`import "k8s.io/kubernetes/pkg/util/mount"`

And then you can do things like:

```Go
mounter := mount.Mounter{}
mountPoints, err := mounter.List()
```

where `mountPoints` is an array of:

```Go
// This represents a single line in /proc/mounts or /etc/fstab.
type MountPoint struct {
	Device string
	Path   string
	Type   string
	Opts   []string
	Freq   int
	Pass   int
}
```

It also provides some other useful routines like `mounter.PathIsDevice()` or `mount.IsNotMountPoint()` that you could use in your previous patch to reduce code duplication.

File: mount.go [hunk 20]
```diff
@@ -148,3 +148,20 @@ func getDevicePathAndFsType(mountPoint string) (devicePath, fsType string, err e
...
+func isDeviceMapper(major, minor int) (bool, error) {
+
+	//Check if /sys/dev/block/${major}-${minor}/dm exists
+	sysPath := fmt.Sprintf("/sys/dev/block/%d:%d/dm", major, minor)
+
+	_, err := os.Stat(sysPath)
+
+	if err == nil {
+		return true, nil
+	} else if os.IsNotExist(err) {
```
Comment (sameo) [inline mount.go]: Why do we make that not an error ?

File: hyperstart.go [hunk 21]
```diff
@@ -403,6 +403,50 @@ func (h *hyper) exec(pod *Pod, c Container, process Process, cmd Cmd) error {
 	return nil
 }
 
+func (h *hyper) addDrives(pod *Pod) error {
```
Comment (sameo) [inline hyperstart.go]: Here as well, this should be a `Pod` private method for other hypervisor implementations to use the same logic.

**Comment by sameo** [comment]
This mostly looks good to me, I only have requests for minor changes.
@amshinde One thing I'd like you to do though is to split this PR in two: One for the bind mount part and another one for the devicemapper support.

File: hyperstart.go (hunk 14 above)
Comment (amshinde) [inline hyperstart.go]: One is for the container rootfs and the other is for the bind-mounts that we created in the shared directory

File: container.go [hunk 22]
```diff
@@ -55,6 +55,12 @@ type Mount struct {
 
 	// fstab style mount options
 	Options []string
+
+	// IgnoreMount denotes these need to be handled explicitly by the agent within the VM
+	IgnoreMount bool
```
Comment (amshinde) [inline container.go]: Will do

File: hyperstart.go (hunk 13 above)
Comment (amshinde) [inline hyperstart.go]: The implementation for the bind-mounting is agent/hyperstart specific. That's why I put it here, along the same lines of bindUnmountContainerRootfs function that was defined for hyperstart

File: mount.go (hunk 18 above)
Comment (amshinde) [inline mount.go]: Thats right, I'll add  example to the function documentation

File: mount.go (hunk 20 above)
Comment (amshinde) [inline mount.go]: We want to check if the path exists or not to determine if its a devicemapper device. If it exists, we return true else false if it does not. All other errors are treated as errors

File: mount.go (hunk 16 above)
Comment (amshinde) [inline mount.go]: @sameo The above code does not seem to give the correct major-minor numbers. I know thats the how its documented in libc. This is what I followed:
https://github.com/golang/go/issues/8106

File: mount.go [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// These mounts need to be created by the agent within the VM
+var systemMounts = []string{"/proc", "/dev", "/dev/pts", "/dev/shm", "/dev/mqueue", "/sys", "/sys/fs/cgroup"}
```
Comment (sboeuf) [inline mount.go]: Should be defined in `pkg/hyperstart` instead.

Comment (sameo) [inline mount.go]: This is not really an hyperstart specific list, it can potentially apply to any other agent.

File: hyperstart.go [hunk 24]
```diff
@@ -220,6 +222,98 @@ func (h *hyper) bindMountContainerRootfs(podID, cID, cRootFs string, readonly bo
...
+			Source:       filename,
+			Path:         m.Destination,
+			ReadOnly:     readonly,
+			DockerVolume: false,
+		}
+
+		fsmap = append(fsmap, fsmapDesc)
+	}
+
+	c.storeMounts()
```
Comment (amshinde) [inline hyperstart.go]: No, we updated the HostPath for the mount here.
I have moved the storeMount outside at the pod start in any case, will push an update

File: utils.go [hunk 25]
```diff
@@ -41,3 +42,22 @@ func fileCopy(srcPath, dstPath string) error {
...
+	}
+
+	return b, nil
+}
+
+func reverseString(s string) string {
+	r := []rune(s)
+	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
+		r[i], r[j] = r[j], r[i]
+	}
```
Comment (amshinde) [inline utils.go]: Fixing

File: hyperstart.go [hunk 26]
```diff
@@ -220,6 +222,98 @@ func (h *hyper) bindMountContainerRootfs(podID, cID, cRootFs string, readonly bo
...
+		filename := fmt.Sprintf("%s-%s-%s", c.id, hex.EncodeToString(randBytes), filepath.Base(m.Destination))
+		mountDest := filepath.Join(defaultSharedDir, podID, filename)
+
+		err = bindMount(m.Source, mountDest, false)
+		if err != nil {
+			return nil, err
+		}
+
+		m.IgnoreMount = false
+		m.HostPath = mountDest
```
Comment (amshinde) [inline hyperstart.go]: Because I am passing a slice here, the underlying elements can be modified.

**Comment by amshinde** [comment]
Closing this PR. Opening 2 others for mounts and devicemapper

File: filesystem.go [hunk 27]
```diff
@@ -570,6 +599,32 @@ func (fs *filesystem) storeContainerProcess(podID, containerID string, process P
...
+	data, err := fs.fetchResource(false, podID, containerID, mountsFileType)
+	if err != nil {
+		return []Mount{}, err
+	}
+
+	switch mounts := data.(type) {
+	case []Mount:
+		return mounts, nil
+	default:
+		return []Mount{}, fmt.Errorf("Unknown mounts type : [%T]", mounts)
```
Comment (amshinde) [inline filesystem.go]: Used this since I wanted to display the type here

Comment (sboeuf) [inline filesystem.go]: You can display the type outside this switch since you assigned the `mounts` variable before entering the switch, right ?

Comment (sboeuf) [inline filesystem.go]: I am sorry I have deleted your comment by mistake. But I agree that you need this way to display the type.

File: mount.go (hunk 19 above)
Comment (amshinde) [inline mount.go]: sameo had a look at that package, there is one function that I can use for parsing /proc/mounts file as of now. I'd rather just do that, then import that package.
